### ISSUE TITLE
[target] Saving .target xml file ignores existing file's line endings

### ISSUE THREAD

**Issue by haubi** [eclipse-pde/eclipse.pde#1059]
So we do have some .target files committed to our Git repo for ages, and we use the Eclipse IDE on both Linux and Windows. 

Recently, I've added Maven artifacts to that .target files, using Eclipse IDE on Linux.

Now, when developers on Windows just perform "Apply" in the Target Platform Preferences dialog, even without actually changing anything there, they see changes in the line endings on all the .target files that reference Maven artifacts.

My **first** thought was for the difference with Maven artifacts:
The result is PR https://github.com/eclipse-m2e/m2e-core/pull/1649, to avoid writing the .target file when the content is identical.

My **second** thought was for the line endings, and this is what I figured out:
It actually is the System.lineSeparator() used, and I fail to find any code in this area to eventually reuse the existing file's line separator, or to look at Eclipse or Git properties (e.g. core.autocrlf) for the preferred line separator.
So even if the .target file is written because of real changes, the line separator may change for the whole file, depending on the OS where Eclipse was running on to create that file before.

After all, it really is confusing and annoying to developers, when they see local changes to files they don't touch at all, and leads to additional effort to revert and/or ignore these changes.

Any thought for how to work around this problem until a proper fix is available?

For the records: The stack in **Eclipse 2023-09** for where the first line separator added to the .target file content is:
```
ToXMLStream.startDocumentInternal() line: 188	
ToXMLStream(SerializerBase).startDocument() line: 1076	
DOM2TO.parse() line: 91	
TransformerImpl.transformIdentity(Source, SerializationHandler) line: 718	
TransformerImpl.transform(Source, SerializationHandler, String) line: 780	
TransformerImpl.transform(Source, Result) line: 395	
TargetDefinitionPersistenceHelper.persistXML(ITargetDefinition, OutputStream) line: 131	
TargetDefinition.write(OutputStream) line: 825	
WorkspaceFileTargetHandle.save(ITargetDefinition) line: 81	
TargetPlatformService.saveTargetDefinition(ITargetDefinition) line: 303	
TargetPlatformPreferencePage.performOk() line: 952	
TargetPlatformPreferencePage(PreferencePage).performApply() line: 457	
TargetPlatformPreferencePage(PreferencePage).lambda$1(SelectionEvent) line: 289	
0x00007f9c88f9c6c0.accept(Object) line: not available	
SelectionListener$1.widgetSelected(SelectionEvent) line: 84	
TypedListener.handleEvent(Event) line: 252	
EventTable.sendEvent(Event) line: 89	
Display.sendEvent(EventTable, Event) line: 5855	
Button(Widget).sendEvent(Event) line: 1529	
Display.runDeferredEvents() line: 5065	
Display.readAndDispatch() line: 4517	
WorkbenchPreferenceDialog(Window).runEventLoop(Shell) line: 823	
WorkbenchPreferenceDialog(Window).open() line: 799	
OpenPreferencesAction.run() line: 66
...
```
Thanks!

**Comment by merks** [comment]
Yes, just last week I was trying to track down some strange behaviors with respect to line endings.  I.e., here a "no-op" change results in line-ending changes in the comments:

![image](https://github.com/eclipse-pde/eclipse.pde/assets/208716/54a60ff5-7095-4e5b-92d6-62282ef5e542)

But then in a debug environment, I could not reproduce the problem.  Very annoying!

I also came to the conclusion that it appears to be using system line ending regardless of any preferences and regardless of the current line ends of the content...

In your case, what are the current line endings on disk before the change?

**Comment by merks** [comment]
This looks to be one of those problems that is so old so as to make it incredible that it still a problem today.

https://stackoverflow.com/questions/14495480/how-to-prevent-xml-transformer-from-changing-line-endings

One would hope that this would work:

![image](https://github.com/eclipse-pde/eclipse.pde/assets/208716/ddb3ae40-6c2e-454f-86de-41243c9d9dd1)

But that property isn't transferred:

![image](https://github.com/eclipse-pde/eclipse.pde/assets/208716/1a68dea4-9549-455d-9d42-af63ea27fa76)

Though the underlying implementation handles that property:

![image](https://github.com/eclipse-pde/eclipse.pde/assets/208716/8ee95ced-6743-4d76-808e-97d60bd7b74b)

Conclusion:  No one really needs to control the line separator produced by a transformer.

Compound this with the fact that such serialization happens the implementation of each location type, it's just not at all trivial to fix once.  So I think this problem would take a significant investment of time to fix across multiple components, and only the people on Windows will care (which does include me).  Therefore, unless someone is willing to invest a significant amount of time, this problem is unlikely to be fixed.  😱 

(Oh yes, and apparently there seems to be no way to set the `line.separator` system property from a command line.)

**Comment by haubi** [comment]
> In your case, what are the current line endings on disk before the change?

Before pressing 'Apply' in the Target Platform Preferences page, it was LF on disk, and CRLF afterwards.
But the other way round likely is possible as well - as soon as .target platform files referencing maven artifacts are loaded into the workspace on both platforms.

> (Oh yes, and apparently there seems to be no way to set the line.separator system property from a command line.)

IIRC, I also have tried and failed to properly set the `-Dline.separator` commandline option - but I don't believe this would be the way to go anyway.

After all, my workaround here is to set both the Git attributes `text` and `eol`:
```
$ cat .gitattributes
*.target text eol
```
`eol` prevents Git from believing the file has changed when line endings change, and `text` prevents Git from committing CRLF at all.
